web3/utils/datatypes.py

```diff
@@ -0,0 +1,45 @@
+from functools import (
+    partial,
+)
+from eth_utils import (
+    to_dict,
+)
+
+from web3.utils.normalizers import (
+    normalize_abi,
+    normalize_address,
+    normalize_bytecode,
+)
+
+
+class PropertyCheckingFactory(type):
+    ens = None
+
+    @to_dict
+    def normalize_properties(attributes, values, ens):
+        normalizers = {
+            'abi': normalize_abi,
+            'address': partial(normalize_address, ens),
+            'bytecode': normalize_bytecode,
+            'bytecode_runtime': normalize_bytecode,
+        }
+
+        for attribute, value in zip(attributes, values):
+            if attribute in normalizers:
+                normalizer = normalizers[attribute]
+                yield attribute, normalizer(value)
+            else:
+                yield attribute, value
+
+    def __new__(mcs, name, bases, namespace):
+        for key in namespace:
+            if not hasattr(bases[0], key):
```

Comment:
*nitpick*

This likely won't cause any issues with the way we're using this, but it should probably check all of the bases or even the set of all `set(concat(base.__mro__ for base in bases))` to cover the full inheritance tree.